specs/MultiProfile.md

```diff
@@ -0,0 +1,274 @@
+                })
+                .Get()));
+    }
+    else
+    {
+        CHECK_FAILURE(webViewEnvironment4->CreateCoreWebView2ControllerWithOptions(
+            m_mainWindow, options.Get(),
+            Callback<ICoreWebView2CreateCoreWebView2ControllerCompletedHandler>(
+                this, &AppWindow::OnCreateCoreWebView2ControllerCompleted)
+                .Get()));
+    }
+
+    return S_OK;
+}
+```
+
+### Access the profile property of WebView2
+
+```cpp
+HRESULT AppWindow::OnCreateCoreWebView2ControllerCompleted(HRESULT result, ICoreWebView2Controller* controller)
+{
+    // ...
+
+    m_controller = controller;
+    
+    // Gets the webview object from controller.
+    wil::com_ptr<ICoreWebView2> coreWebView2;
+    CHECK_FAILURE(m_controller->get_CoreWebView2(&coreWebView2));
+
+    auto webview7 = coreWebView2.try_query<ICoreWebView2_7>();
+    if (webview7)
+    {
+        // Gets the profile property of webview.
+        wil::com_ptr<ICoreWebView2Profile> profile;
+        CHECK_FAILURE(webview7->get_Profile(&profile));
+
+        // Accesses the profile object.
+        BOOL inPrivateModeEnabled = FALSE;
+        CHECK_FAILURE(profile->get_IsInPrivateModeEnabled(&inPrivateModeEnabled));
+        wil::unique_cotaskmem_string profile_path;
+        CHECK_FAILURE(profile->get_ProfilePath(&profile_path));
+        std::wstring str(profile_path.get());
+        m_profileDirName = str.substr(str.find_last_of(L'\\') + 1);
+        
+        // update window title with m_profileDirName
+        UpdateAppTitle();
+
+        // update window icon
+        SetAppIcon(inPrivate);        
+    }
+  
+    // ...
+}
+```
+## .NET and WinRT
+
+### Create WebView2 with a specific profile, then access the profile property of WebView2
+
+```csharp
+CoreWebView2Environment _webViewEnvironment;
+public CreateWebView2ControllerWithOptions(IntPtr parentHWND, string profileName, bool isInPrivate)
+{
+    CoreWebView2ControllerOptions options = _webViewEnvironment.CreateCoreWebView2ControllerOptions(profileName, isInPrivate);
+    CoreWebView2Controller webView2Controller = await _webViewEnvironment.CreateCoreWebView2ControllerWithOptionsAsync(parentHWND, options);
+    string profilePath = webView2Controller.CoreWebView2.Profile.ProfilePath;
+    string profileDirName = profilePath.Substring(profilePath.LastIndexOf('\\') + 1);
+    bool inPrivate = webView2Controller.CoreWebView2.Profile.IsInPrivateModeEnabled;
+
+    // update window title with profileDirName
+    UpdateAppTitle();
+
+    // update window icon
+    SetAppIcon(inPrivate);
+}
+```
+
+# API Details
+
+## Win32 C++
+
+```IDL
+interface ICoreWebView2ControllerOptions;
+interface ICoreWebView2Environment5;
+interface ICoreWebView2_7;
+interface ICoreWebView2Profile;
+
+[uuid(C2669A3A-03A9-45E9-97EA-03CD55E5DC03), object, pointer_default(unique)]
+interface ICoreWebView2ControllerOptions : IUnknown {
```

Comment:
Is there a point at which we'll convert the interface to MIDL3, so you don't have to dual-author the C++ and C# interop types? We'll have `Microsoft.Windows.UI.Core.WindowId` at some point so you don't have to work around HWND missing in WinRT metadata.

Reply:
TLDR: In the long term yes I'd like to make our IDL3 the source of truth and from it generate our .NET & COM APIs or replace those with just WinRT

This is our COM IDL(1) style IDL file to define our COM APIs usable down to Win7. Because it goes to Win7 where we don't have various aspects of WinRT available the conventions of the COM API are similar to but are not the same as the WinRT ABI. We later added WinRT support in part by generating a MIDL3 from our COM IDL in a tool that understands how to translate between our conventions and WinRT. In the future I'd like to make MIDL3 the source of truth from which we generate our COM IDL and then after that if usage drops far enough deprecating in favor of COM ABI WinRT.